trusted/pal/src/pal.rs

```diff
@@ -0,0 +1,174 @@
+    ///
+    /// # Returns
+    ///
+    /// An error if the signature cannot be verifed or is invalid, a success otherwise.
+    fn verify(&self, data: &[u8], signature: &[u8]) -> Result<()>;
+
+    /// Gets serialized public key used for signing by the trusted application.
+    /// The signing key is generated at the start of the trusted application and
+    /// remains unchanged through the lifetime.
+    ///
+    /// # Note
+    ///
+    /// Public signing key can be used to derive a trusted application identity.
+    /// For example, a trusted application that represents a node in Raft cluster
+    /// running inside of a group trusted hosts, it is important that Raft node
+    /// identity cannot be forged, does not require coordination to pick one and
+    /// has low collision chance. Hash of the public signing key is an identity
+    /// mechanism that is compliant with these requirements.
+    fn public_signing_key(&self) -> Vec<u8>;
+}
+
+/// Represents a trusted host, abstracting away the details of how trusted
+/// application is hosted (e.g. trusted host can be restricted kernel bare
+/// metal based encrypted virtual machine or a linux kernel docker based
+/// encrypted virtual machine).
+pub trait Host {
+    /// Gets attestation that represens a set of claims about the trusted application
+    /// and its host. The attestation is generated after the trusted application and
+    /// host are fully initialiezed, the attestation doesn't change throughout the
+    /// lifetime.
+    fn get_self_attestation(&self) -> Box<dyn Attestation>;
+
+    /// Gets serialized configuration that stays immutable through the lifetime of
+    /// the trusted application.
+    fn get_self_config(&self) -> Vec<u8>;
+
+    /// Sends messages through the communication channel that connects the trusted
+    /// application to the untrusted launcher.
+    ///
+    /// # Argumnets
+    ///
+    /// * `messages` - A set of messsages to send through the channel.
+    ///
+    /// # Returns
+    ///
+    /// Error if the communication channel is irrepairably broken, a success otherise.
+    fn send_messages(&mut self, messages: &[Message]) -> Result<()>;
+
+    /// Attempts to deserialize peer attestation and perform initial verification
+    /// of the attestation. The application specific verification (e.g. ensuring
+    /// that attestation points to a particular binary and configuration) is the
+    /// responsibility of the trusted application.
+    ///
+    /// # Arguments
+    ///
+    /// * `peer_attestation` - A serialized attestation of a peer trusted application.
+    ///
+    /// # Returns
+    ///
+    /// A deserialized and initially verified peer attestation if success, an error
+    /// otherwise. An error may be caused by malformed serialized representation or
+    /// by failing verification.
+    fn verify_peer_attestation(&self, peer_attestation: &[u8]) -> Result<Box<dyn Attestation>>;
```

Comment:
As per previous comment, `AttestationVerifier` could be a separate entity, since it doesn't need anything from the Restricted Kernel. You just provide it with a policy (which itself could be generated from self evidence, i.e. policy = is a given evidence equal to self evidence) and it verifies it.